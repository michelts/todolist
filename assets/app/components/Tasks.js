const Tasks = () => null;

export default Tasks;
