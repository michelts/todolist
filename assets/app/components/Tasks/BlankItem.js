import React from 'react';
import ListGroup from 'react-bootstrap/ListGroup';

const BlankItem = () => (
  <ListGroup.Item>
    All good! You are done with your tasks :)
  </ListGroup.Item>
);

export default BlankItem;
