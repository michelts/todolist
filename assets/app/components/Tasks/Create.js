import React from 'react';
import axios from 'axios';
import Button from 'react-bootstrap/Button';

const CreateButton = ({ tasks, setTasks }) => {
  const [creating, setCreating] = React.useState(false);

  const handleClick = () => {
    setCreating(true);
    axios.post('/api/v1/tasks/', { description: '' }).then(({ data }) => {
      setTasks((state) => state.set(data.id, data));
      setCreating(false);
    });
  };

  return (
    <Button
      disabled={tasks === undefined || creating}
      onClick={handleClick}
    >
      Add new task
    </Button>
  );
};

export default CreateButton;
