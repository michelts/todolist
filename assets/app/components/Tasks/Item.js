import React from 'react';
import axios from 'axios';
import ListGroup from 'react-bootstrap/ListGroup';
import Form from 'react-bootstrap/Form';
import Button from 'react-bootstrap/Button';
import EditText from 'common/components/EditText';

const TaskItem = ({
  task: {
    id,
    ...task
  },
  setTasks,
}) => {
  const priorityRef = React.useRef();

  const handleSaveDescription = React.useCallback((value) => {
    const payload = { ...task, description: value };
    axios.put(`/api/v1/tasks/${id}/`, payload).then(({ data }) => {
      setTasks((state) => state.set(id, data));
    });
  }, [id, setTasks]);

  const handleSaveDueDate = React.useCallback((value) => {
    const payload = { ...task, due_date: value };
    axios.put(`/api/v1/tasks/${id}/`, payload).then(({ data }) => {
      setTasks((state) => state.set(id, data));
    });
  }, [id, setTasks]);

  const priorities = [0, 1, 2, 3, 4];

  const handlePriorityChange = React.useCallback(() => {
    const { current: { value } } = priorityRef;
    const payload = { ...task, priority: parseInt(value, 10) };
    axios.put(`/api/v1/tasks/${id}/`, payload).then(({ data }) => {
      setTasks((state) => state.set(id, data));
    });
  }, [id, setTasks]);

  const handleCompleted = React.useCallback(() => {
    const payload = { ...task, is_completed: !task.is_completed };
    axios.put(`/api/v1/tasks/${id}/`, payload).then(({ data }) => {
      setTasks((state) => state.set(id, data));
    });
  }, [id, setTasks]);

  const handleArchive = React.useCallback(() => {
    axios.delete(`/api/v1/tasks/${id}/`).then(() => {
      setTasks((state) => state.delete(id));
    });
  }, [id, setTasks]);

  return (
    <ListGroup.Item>
      <div className="d-flex justify-content-between align-items-center">
        <div className="w-10 text-center">
          <div>Done?</div>
          <Form.Check
            name="is-completed"
            type="checkbox"
            checked={task.is_completed}
            onClick={handleCompleted}
          />
        </div>
        <div className="flex-grow-1 ml-4">
          <EditText
            data-name="description"
            value={task.description || 'Fill with the task description'}
            onSave={handleSaveDescription}
          />
          <span className="text-muted">
            <EditText
              data-name="due-date"
              value={task.due_date || 'No date'}
              onSave={handleSaveDueDate}
            />
          </span>
        </div>
        <div className="w-10 ml-4 text-center">
          <div className="mb-1">Priority</div>
          <Form.Control
            as="select"
            name="priority"
            size="sm"
            custom
            ref={priorityRef}
            onChange={handlePriorityChange}
          >
            {priorities.map((index) => (
              <option
                key={index}
                value={index}
                selected={index === task.priority}
              >
                {index}
              </option>
            ))}
          </Form.Control>
        </div>
        <div className="w-10 ml-4">
          <Button
            name="archive"
            variant="danger"
            className="btn-sm"
            onClick={handleArchive}
          >
            Archive
          </Button>
        </div>
      </div>
    </ListGroup.Item>
  );
};

export default TaskItem;
