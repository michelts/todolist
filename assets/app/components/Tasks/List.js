import React from 'react';
import ListGroup from 'react-bootstrap/ListGroup';

const TaskList = ({ children }) => (
  <ListGroup className="mb-3">
    {children}
  </ListGroup>
);

export default TaskList;
