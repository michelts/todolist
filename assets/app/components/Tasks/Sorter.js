import React from 'react';
import axios from 'axios';
import { OrderedMap } from 'immutable';
import Form from 'react-bootstrap/Form';

const Sorter = ({ setTasks, disabled }) => {
  const inputRef = React.useRef();
  const [timeoutId, setTimeoutId] = React.useState(null);

  const handleChange = React.useCallback(() => {
    const { current: { value } } = inputRef;

    if (timeoutId !== null) {
      clearTimeout(timeoutId);
    }

    setTimeoutId(
      setTimeout(() => {
        setTasks(undefined);
        axios.get(`/api/v1/tasks/?sort=${value}`).then(({ data }) => {
          setTasks(
            OrderedMap(data.map((task) => [task.id, task])),
          );
        });
      }, 1000),
    );
  }, [inputRef, setTasks]);

  return (
    <Form.Control
      as="select"
      size="sm"
      custom
      ref={inputRef}
      disabled={disabled}
      onChange={handleChange}
    >
      <option value="due_date">Sort by date</option>
      <option value="priority">Sort by priority</option>
    </Form.Control>
  );
};

export default Sorter;
