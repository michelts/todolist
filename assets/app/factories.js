import Factory from 'rosie';

const UserFactory = new Factory()
  .sequence('id', (index) => index + 1)
  .sequence('username', (index) => `username_${index}`)
  .sequence('name', (index) => `Name-${index}`);

export {
  UserFactory,
};
