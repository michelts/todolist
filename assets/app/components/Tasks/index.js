import React from 'react';
import Button from 'react-bootstrap/Button';
import ListGroup from 'react-bootstrap/ListGroup';
import Sorter from './Sorter';
import Item from './Item';

const Tasks = () => (
  <>
    <div className="d-flex justify-content-start align-items-center">
      <div>
        <h1>Tasks</h1>
      </div>
      <div className="ml-3">
        <Sorter />
      </div>
    </div>
    <ListGroup className="mb-3">
      <Item item={{ description: 'Cras justo odio', due_date: null }} />
      <Item item={{ description: 'Lorem Ipsum dolor sit amet', due_date: '2019-12-29' }} />
      <Item item={{ description: 'Cras justo odio', due_date: null }} />
      <Item item={{ description: 'Lorem Ipsum dolor sit amet', due_date: '2019-12-29' }} />
    </ListGroup>
    <Button>Add new</Button>
  </>
);

export default Tasks;
