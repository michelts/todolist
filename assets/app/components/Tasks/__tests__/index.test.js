import React from 'react';
import axios from 'axios';
import { shallow } from 'enzyme';
import { TaskFactory } from '../../../factories';
import Tasks from '../index';

const getWrapper = () => {
  const wrapper = shallow(<Tasks />);
  return { wrapper };
};

describe('Tasks component', () => {
  beforeEach(() => {
    jest.spyOn(React, 'useEffect').mockImplementation(effect => effect());
  });

  it('should render a title, disabled create and sort controls, and the list as loading while list is undefined', () => {
    axios.get.mockResolvedValue({});
    const { wrapper } = getWrapper();
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a title, create and sort controls, and the list of tasks when available', async () => {
    axios.get.mockResolvedValue({ data: TaskFactory.buildList(1) });
    const { wrapper } = getWrapper();
    await axios.get;
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a title, create and sort controls, and an empty list when tasks are empty', async () => {
    axios.get.mockResolvedValue({ data: [] });
    const { wrapper } = getWrapper();
    await axios.get;
    expect(wrapper).toMatchSnapshot();
  });
});
