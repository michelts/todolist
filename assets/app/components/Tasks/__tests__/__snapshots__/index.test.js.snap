// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tasks component should render a title, create and sort controls, and an empty list when tasks are empty 1`] = `
<Fragment>
  <div
    className="d-flex justify-content-start align-items-center"
  >
    <div>
      <h1>
        Tasks
      </h1>
    </div>
    <div
      className="ml-3"
    >
      <Sorter
        disabled={false}
        setTasks={[Function]}
      />
    </div>
  </div>
  <ListGroup
    className="mb-3"
    horizontal={null}
    variant={null}
  >
    <BlankItem />
  </ListGroup>
  <CreateButton
    disabled={false}
    setTasks={[Function]}
  >
    Add new task
  </CreateButton>
</Fragment>
`;

exports[`Tasks component should render a title, create and sort controls, and the list of tasks when available 1`] = `
<Fragment>
  <div
    className="d-flex justify-content-start align-items-center"
  >
    <div>
      <h1>
        Tasks
      </h1>
    </div>
    <div
      className="ml-3"
    >
      <Sorter
        disabled={false}
        setTasks={[Function]}
      />
    </div>
  </div>
  <ListGroup
    className="mb-3"
    horizontal={null}
    variant={null}
  >
    <TaskItem
      key="2"
      setTasks={[Function]}
      task={
        Object {
          "description": "Task 1",
          "due_date": "2020-05-22",
          "id": 2,
          "is_completed": false,
          "priority": 0,
          "selected": false,
        }
      }
    />
  </ListGroup>
  <CreateButton
    disabled={false}
    setTasks={[Function]}
  >
    Add new task
  </CreateButton>
</Fragment>
`;

exports[`Tasks component should render a title, disabled create and sort controls, and the list as loading while list is undefined 1`] = `
<Fragment>
  <div
    className="d-flex justify-content-start align-items-center"
  >
    <div>
      <h1>
        Tasks
      </h1>
    </div>
    <div
      className="ml-3"
    >
      <Sorter
        disabled={true}
        setTasks={[Function]}
      />
    </div>
  </div>
  <ListGroup
    className="mb-3"
    horizontal={null}
    variant={null}
  >
    <Loading />
  </ListGroup>
  <CreateButton
    disabled={true}
    setTasks={[Function]}
  >
    Add new task
  </CreateButton>
</Fragment>
`;
